src/main/java/org/opentripplanner/raptor/rangeraptor/TimeBasedRoutingStrategy.java

```diff
@@ -0,0 +1,249 @@
+package org.opentripplanner.raptor.rangeraptor;
+
+import java.util.function.IntConsumer;
+import javax.annotation.Nonnull;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoundProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoutingStrategy;
+import org.opentripplanner.raptor.rangeraptor.internalapi.SlackProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.WorkerLifeCycle;
+import org.opentripplanner.raptor.rangeraptor.transit.TransitCalculator;
+import org.opentripplanner.raptor.spi.RaptorAccessEgress;
+import org.opentripplanner.raptor.spi.RaptorConstrainedTripScheduleBoardingSearch;
+import org.opentripplanner.raptor.spi.RaptorTimeTable;
+import org.opentripplanner.raptor.spi.RaptorTripSchedule;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleBoardOrAlightEvent;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleSearch;
+import org.opentripplanner.raptor.spi.TransitArrival;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.request.TripScheduleBoardSearch;
+
+/**
+ * This class contains code which is shared by all time-dependent routing strategies. It also
+ * defines abstract methods, which the individual strategies must implement.
+ */
+public abstract class TimeBasedRoutingStrategy<T extends RaptorTripSchedule>
+  implements RoutingStrategy<T> {
+
+  protected final SlackProvider slackProvider;
+  protected final TransitCalculator<T> calculator;
+  private final RoundProvider roundProvider;
+  private boolean inFirstIteration = true;
+  private boolean hasTimeDependentAccess = false;
+  private RaptorTimeTable<T> timeTable;
+  private RaptorTripScheduleSearch<T> tripSearch;
+
+  protected TimeBasedRoutingStrategy(
+    SlackProvider slackProvider,
+    TransitCalculator<T> calculator,
+    RoundProvider roundProvider,
+    WorkerLifeCycle subscriptions
+  ) {
+    this.slackProvider = slackProvider;
```

Comment:
From a quick check, it seems that all strategy have a private state member.
Should it be moved to parent class?

Reply:
Not relevant any more. This is refactored to use composition, and not inheritance.